cli/util/upload_state.go

```diff
@@ -0,0 +1,209 @@
+		return err
+	}, maxRetries, debug)
+	if err != nil {
+		return cli.Exit(fmt.Errorf("failed to get container with ID %s: %w", containerID, err), 1)
+	}
+	containerMagic := containerObj.Attribute("Magic")
+
+	logger := zap.NewExample()
+	cfg, err := options.GetConfigFromContext(ctx)
+	if err != nil {
+		return cli.Exit(err, 1)
+	}
+
+	chain, store, err := server.InitBlockChain(cfg, logger)
+	if err != nil {
+		return cli.Exit(err, 1)
+	}
+
+	magic := strconv.Itoa(int(chain.GetConfig().Magic))
+	if containerMagic != magic {
+		return cli.Exit(fmt.Sprintf("container magic %s does not match the network magic %s", containerMagic, magic), 1)
+	}
+	syncInterval := uint32(chain.GetConfig().ProtocolConfiguration.StateSyncInterval)
+	stateObjCount, err := searchStateIndex(ctx, pWrapper, containerID, acc.PrivateKey(), attr, uint(syncInterval), maxRetries, debug)
+
+	stateModule := chain.GetStateModule()
+	currentHeight := stateModule.CurrentLocalHeight()
+
+	currentStateIndex := currentHeight / syncInterval
+	if currentStateIndex <= stateObjCount {
+		fmt.Fprintf(ctx.App.Writer, "No new states to upload. Last state sync at block %d, current height %d\n",
+			stateObjCount*syncInterval, currentHeight)
+		return nil
+	}
+
+	fmt.Fprintf(ctx.App.Writer, "Latest state object found at height %d, current height %d, %d new states to upload\n", stateObjCount*syncInterval, currentHeight, currentStateIndex-stateObjCount)
+
+	for state := stateObjCount; state < currentStateIndex; state++ {
+		stateRoot, err := stateModule.GetStateRoot(state * syncInterval)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to get state root at height %d: %v", state, err), 1)
+		}
+
+		nodes, err := traverseRawMPT(stateRoot.Root, store, mpt.ModeLatest)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to traverse MPT at height %d: %v", state*syncInterval, err), 1)
+		}
+
+		objBytes, err := encodeMPTNodes(nodes)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to encode MPT nodes at height %d: %v", state*syncInterval, err), 1)
+		}
+
+		h, err := chain.GetHeader(chain.GetHeaderHash(state * syncInterval))
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to get header at height %d: %v", state*syncInterval, err), 1)
+		}
+
+		attrs := []object.Attribute{
+			*object.NewAttribute(attr, strconv.Itoa(int(state*syncInterval))),
+			*object.NewAttribute("Timestamp", strconv.FormatInt(time.Now().Unix(), 10)),
+			*object.NewAttribute("StateRoot", stateRoot.Root.StringLE()),
+			*object.NewAttribute("StateSyncInterval", strconv.Itoa(int(syncInterval))),
+			*object.NewAttribute("BlockTime", strconv.FormatUint(h.Timestamp, 10)),
+		}
+		_, err = uploadObj(ctx.Context, pWrapper, signer, containerID, objBytes, attrs)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to upload object at height %d: %v", state*syncInterval, err), 1)
+		}
+		fmt.Fprintf(ctx.App.Writer, "Uploaded state object at height %d\n", state*syncInterval)
+	}
+	err = store.Close()
```

Comment:
It must be a deferred statement. Any error above will give you an improperly closed storage.